src/freenet/node/diagnostics/threads/DefaultThreadDiagnostics.java

```diff
@@ -0,0 +1,122 @@
+/* This code is part of Freenet. It is distributed under the GNU General
+ * Public License, version 2 (or at your option any later version). See
+ * http://www.gnu.org/ for further details of the GPL. */
+package freenet.node.diagnostics.threads;
+
+import freenet.node.*;
+import freenet.node.diagnostics.*;
+import freenet.support.*;
+
+import java.lang.management.*;
+import java.util.*;
+import java.util.concurrent.atomic.*;
+import java.util.stream.*;
+
+/**
+ * Runnable thread to retrieve node thread's information and compiling it into
+ * an array of NodeThreadInfo objects.
+ */
+public class DefaultThreadDiagnostics implements Runnable, ThreadDiagnostics {
+    private final String name;
+    private final int monitorInterval;
+
+    private final NodeStats nodeStats;
+    private final Ticker ticker;
+
+    /** Sleep interval to calculate % CPU used by each thread */
+    private static final int MONITOR_INTERVAL = 1000;
+    private static final String MONITOR_THREAD_NAME = "NodeDiagnostics: thread monitor";
+
+    private final AtomicReference<List<NodeThreadInfo>> nodeThreadInfo = new AtomicReference<>(new ArrayList<>());
+
+    private final ThreadMXBean threadMxBean = ManagementFactory.getThreadMXBean();
+
+    /**
+     * @param nodeStats Used to retrieve data points
+     * @param ticker Used to queue timed jobs
+     * @param name Thread name
+     * @param monitorInterval Sleep intervals to retrieve CPU usage
+     */
+    public DefaultThreadDiagnostics(NodeStats nodeStats, Ticker ticker, String name, int monitorInterval) {
+        this.nodeStats = nodeStats;
+        this.ticker = ticker;
+        this.name = name;
+        this.monitorInterval = monitorInterval;
+    }
+
+    /*
+     * @param nodeStats Used to retrieve data points
+     * @param ticker Used to queue timed jobs
+     */
+    public DefaultThreadDiagnostics(NodeStats nodeStats, Ticker ticker) {
+        this(nodeStats, ticker, MONITOR_THREAD_NAME, MONITOR_INTERVAL);
+    }
+
+    private void scheduleNext(int interval) {
+        ticker.queueTimedJob(
+            this,
+            name,
+            interval,
+            false,
+            true
+        );
+    }
+
+    public void start() {
+        scheduleNext(0);
+    }
+
+    private void scheduleNext() {
+        scheduleNext(monitorInterval);
+    }
+
+    private final HashMap<Long, Double> threadCpu = new HashMap<>();
```

Comment:
I think we might need a cleanup for this map at the end of the loop, just something simply like remove all thread IDs that are not in the current threads anymore.

Reply:
Yeah, I was wondering how much could this keep on growing.

Reply:
Well, it would probably be fine for a couple of days, maybe even weeks but it very much depends on how much thread churn there is which without further investigation is hard to determine so I think it’s better to just clean it up. Put in a nice little method called “removeOldThreadDeltas” or something like that and it’s even self-documenting! \o/

Reply:
Well, I could as well do `threadCpu.keySet().removeIf(key -> threads.contains(key) )`. I haven't tested it though.

Reply:
Yes, if that works it would probably be clear enough that a separate method wouldn’t be warranted.

Reply:
Finally added a new private method for it, using streams though.